django/contrib/admin/options.py

```diff
@@ -1583,6 +1584,22 @@ def add_view(self, request, form_url='', extra_context=None):
     def change_view(self, request, object_id, form_url='', extra_context=None):
         return self.changeform_view(request, object_id, form_url, extra_context)
 
+    def get_edited_object_pks(self, request, prefix):
+        pk_pattern = re.compile('{prefix}-\d+-{pk_name}$'.format(prefix=prefix, pk_name=self.model._meta.pk.name))
+        return [
+            value for key, value in request.POST.items() if pk_pattern.match(key)
+        ]
+
+    def get_list_editable_queryset(self, request, prefix):
+        object_pks = self.get_edited_object_pks(request, prefix)
+        modified_objects = self.get_queryset(request).filter(pk__in=object_pks)
+        try:
+            # Force evaluate queryset to verify that pks are valid
+            list(modified_objects)
+        except ValidationError:
+            raise ValidationError("Invalid Primary Key Provided")
```

Comment:
Won't this also result in a 500?

Reply:
Ah yes. It no doubt will. (That's too much DRF that is. 🙂) We need to handle this. 👍

Reply:
@charettes' earlier point:

>  I guess we could try cleaning the submitted pks and and disable the optimization if an exception is raised instead of crashing?

If we here `return self.get_queryset(request)` we're no worse off than currently. 

@AdamDonna Do you think you could put that in and adjust the test(s) for that?

Reply:
I'll add that in and adjust the tests. I think this is a good compromise since it's no worse than it is currently, but most cases will be handled in a better way.